source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,538 @@
+    if (!dispatch_loop) {
+        s_scheduled_service_entry_destroy(entry);
+        return;
+    }
+
+    if (!begin_iteration(entry)) {
+        s_scheduled_service_entry_destroy(entry);
+        return;
+    }
+
+    aws_event_loop_register_tick_start(dispatch_loop->base_loop);
+    // run the full iteration here: local cross-thread tasks
+
+    while (!aws_linked_list_empty(&dispatch_loop->local_cross_thread_tasks)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&dispatch_loop->local_cross_thread_tasks);
+        struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+
+        /* Timestamp 0 is used to denote "now" tasks */
+        if (task->timestamp == 0) {
+            aws_task_scheduler_schedule_now(&dispatch_loop->scheduler, task);
+        } else {
+            aws_task_scheduler_schedule_future(&dispatch_loop->scheduler, task, task->timestamp);
+        }
+    }
+
+    aws_mutex_lock(&dispatch_loop->synced_thread_data.thread_data_lock);
+    dispatch_loop->synced_thread_data.current_thread_id = aws_thread_current_thread_id();
+    dispatch_loop->synced_thread_data.is_executing = true;
+    aws_mutex_unlock(&dispatch_loop->synced_thread_data.thread_data_lock);
+
+    // run all scheduled tasks
+    uint64_t now_ns = 0;
+    aws_event_loop_current_clock_time(dispatch_loop->base_loop, &now_ns);
+    aws_task_scheduler_run_all(&dispatch_loop->scheduler, now_ns);
+    aws_event_loop_register_tick_end(dispatch_loop->base_loop);
+
+    aws_mutex_lock(&dispatch_loop->synced_thread_data.thread_data_lock);
+    dispatch_loop->synced_thread_data.is_executing = false;
+    aws_mutex_unlock(&dispatch_loop->synced_thread_data.thread_data_lock);
+
+    end_iteration(entry);
+}
+
+/**
+ * Checks if a new iteration task needs to be scheduled, given a target timestamp. If so, submits an iteration task to
+ * dispatch queue and registers the pending execution in the event loop's list of scheduled iterations.
+ *
+ * If timestamp==0, the function will always schedule a new iteration as long as the event loop is not suspended.
+ *
+ * The function should be wrapped with dispatch_loop->synced_task_data->lock
+ */
+static void s_try_schedule_new_iteration(struct dispatch_loop_context *dispatch_loop_context, uint64_t timestamp) {
+    struct dispatch_loop *dispatch_loop = dispatch_loop_context->io_dispatch_loop;
+    if (!dispatch_loop || dispatch_loop->synced_task_data.suspended)
+        return;
+    if (!s_should_schedule_iteration(
+            &dispatch_loop->synced_task_data.context->scheduling_state.scheduled_services, timestamp)) {
+        return;
+    }
+    struct scheduled_service_entry *entry = s_scheduled_service_entry_new(dispatch_loop_context, timestamp);
+    aws_linked_list_push_front(
+        &dispatch_loop->synced_task_data.context->scheduling_state.scheduled_services, &entry->node);
+    dispatch_async_f(dispatch_loop->dispatch_queue, entry, s_run_iteration);
```

Comment:
This doesn't seem to take into account that timestamp might be significantly in the future.  I don't get any search hits for `dispatch_after`